Framework/Algorithms/src/CalculateCarpenterSampleCorrection.cpp

```diff
@@ -0,0 +1,467 @@
+  **/
+  const double sigsr = sigir;
+
+  const double delta = COEFF4 * sigir + COEFF5 * sigir * sigir;
+  const double deltp = (delta * sigsct) / (sigsct + sigabs);
+
+  double temp = AttFac(sigir, sigsr, Z);
+  return (deltp / temp);
+}
+
+} // namespace
+
+/**
+ *  This method will change the values in the y_val array to correct for
+ *  multiple scattering absorption. Parameter total_path is in meters, and
+ *  the sample radius is in cm.
+ *
+ *  @param angle_deg ::   The scattering angle (two theta) in degrees
+ *  @param radius ::      The sample rod radius in cm
+ *  @param coeff1 ::      The absorption cross section / 1.81
+ *  @param coeff2 ::      The density
+ *  @param coeff3 ::      The total scattering cross section
+ *  @param wavelength ::          Array of wavelengths at bin boundaries
+    *                     (or bin centers) for the spectrum, in Angstroms
+ *  @param y_val ::       The spectrum values
+ */
+void CalculateCarpenterSampleCorrection::calculate_abs_correction(
+    const double angle_deg, const double radius, const double coeff1,
+    const double coeff2, const double coeff3, const Points &wavelength,
+    HistogramY &y_val) {
+
+  const size_t NUM_Y = y_val.size();
+  bool is_histogram = false;
+  if (wavelength.size() == NUM_Y + 1)
+    is_histogram = true;
+  else if (wavelength.size() == NUM_Y)
+    is_histogram = false;
+  else
+    throw std::runtime_error("Data is neither historgram or density");
+
+  // initialize Z array for this angle
+  vector<double> Z = createZ(angle_deg);
+
+  const double Q2 = coeff1 * coeff2;
+  const double sigsct = coeff2 * coeff3;
+
+  for (size_t j = 0; j < NUM_Y; j++) {
+    double wl_val = wavelength[j];
+    if (is_histogram) // average with next value
+      wl_val = .5 * (wl_val + wavelength[j + 1]);
+
+    y_val[j] = calculate_abs_factor(radius, Q2, sigsct, Z, wl_val);
+  }
+}
+
+void CalculateCarpenterSampleCorrection::calculate_ms_correction(
+    const double angle_deg, const double radius, const double coeff1,
+    const double coeff2, const double coeff3, const Points &wavelength,
+    HistogramY &y_val) {
+
+  const size_t NUM_Y = y_val.size();
+  bool is_histogram = false;
+  if (wavelength.size() == NUM_Y + 1)
+    is_histogram = true;
+  else if (wavelength.size() == NUM_Y)
+    is_histogram = false;
+  else
+    throw std::runtime_error("Data is neither historgram or density");
+
+  // initialize Z array for this angle
+  vector<double> Z = createZ(angle_deg);
+
+  const double Q2 = coeff1 * coeff2;
+  const double sigsct = coeff2 * coeff3;
+
+  for (size_t j = 0; j < NUM_Y; j++) {
+    double wl_val = wavelength[j];
+    if (is_histogram) // average with next value
+      wl_val = .5 * (wl_val + wavelength[j + 1]);
+
+    y_val[j] = calculate_ms_factor(radius, Q2, sigsct, Z, wl_val);
+  }
+}
+
+MatrixWorkspace_sptr CalculateCarpenterSampleCorrection::createOutputWorkspace(
+    const MatrixWorkspace_sptr inputWksp, const std::string ylabel) const {
```

Comment:
It's good practice to use `const MatrixWorkspace_sptr &inputWksp` and take the `shared_ptr` by reference to avoid incremeting the reference count.